### PR TITLE
[data, model, training] fix: Stabilize Qwen3-VL packed SFT

### PR DESCRIPTION
**Pull request by wplf**
# What does this PR do ?

Fixes several Qwen3-VL packed SFT issues that can corrupt MRoPE position ids, route padding tokens through MoE experts, or swap dist-train vision/deepstack tensors. These issues are most visible with packed THD training, mixed image/text-only VLM batches, MoE, and dist-train vision-module splitting.

# Bugfix Summary

- **Packed THD MRoPE**: compute MRoPE from the padded packed sequence layout, temporarily unpack THD input ids to BSHD when needed, then repack position ids back to THD. This avoids position-id/layout mismatch for packed Qwen3-VL forward.
- **MoE padding mask**: preserve the dataset 2D padding mask through Qwen3-VL batch packing, model forward, text model, and decoder block calls. MoE now receives the real padding mask first and only falls back to `lm_input_ids.eq(0)` when no mask exists.
- **Qwen VLM collate mask preservation**: when a batch mixes image examples and text-only examples, keep and merge each processor `attention_mask` alongside `input_ids`. This lets packed SFT derive `padding_mask` and `moe_padding_mask` from real padding metadata instead of token id heuristics.
- **Dist-train vision payload order**: pack deepstack feature chunks first and final `vision_embeds` last, matching `set_dist_train_input_tensors`, which unpacks `chunks[:-1]` as deepstack and `chunks[-1]` as final vision embeddings.
- **Pure-text/deepstack handling**: allow `deepstack_visual_embeds=None` in `split_deepstack_embs` so pure-text microbatches do not crash.

# Validation Findings

## THD MRoPE probe, micro-batch size 3

I added a tiny local probe that feeds already-packed THD `input_ids` with shape `[1, 18]`, `micro_batch_size=3`, `seq_len=6`, and `cu_seqlens=[0, 6, 12, 18]`. The probe stubs the language model and inspects the exact `position_ids` passed from `Qwen3VLModel.forward` into the LM.

With this PR:

```text
rope input shape seen by get_rope_index: [3, 6]
position_ids: 0 1 2 3 4 5 | 0 1 2 3 4 5 | 0 1 2 3 4 5
```

Without this PR, at base `39b79eb78`:

```text
rope input shape seen by get_rope_index: [1, 18]
position_ids: 0 1 2 3 4 5 | 6 7 8 9 10 11 | 12 13 14 15 16 17
```

So the fix makes packed THD MRoPE reset per packed sample using `cu_seqlens`; without it, the flat THD row receives one continuous RoPE index across the whole microbatch.

## Zero-LR MoE BSHD/THD parity check

I also ran a fake-small Qwen3-VL MoE model with padded input, `micro_batch_size=2`, 4 experts, top-2 routing, and `lr=0`, comparing BSHD against THD while holding weights fixed.

```text
steps: 20
max_abs_loss_diff: 0.0
max_abs_forward_diff: 0.0
max_active_abs_forward_diff: 0.0
one-step global_max_abs_grad_diff: 2.2351741790771484e-08
```

The tiny gradient delta is float32 roundoff scale from different layout/reduction order, while forward output and loss are exactly aligned for the same padded tokens and MoE padding mask.

# Tests

- Added focused unit coverage for packed THD MRoPE, explicit packed THD MoE padding-mask propagation, Qwen collate attention-mask preservation, dist-train vision payload order, pure-text deepstack handling, and text-model padding-mask passthrough.

Local checks run:

- `uv run pre-commit run --all-files`
- `uv --project /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge run python -m pytest /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge/tests/unit_tests/models/qwen_vl/modelling_qwen3_vl/test_model.py -k "packed_thd" -q`
- `uv --project /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge run python -m pytest /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge/tests/unit_tests/data/vlm_datasets/test_collate.py -k "qwen2_5_collate_fn" -q`
- `uv --project /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge run python -m pytest /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge/tests/unit_tests/models/qwen_vl/modelling_qwen3_vl/test_text_model_forward.py -q`
- `uv --project /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge run ruff check /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge/tests/unit_tests/models/qwen_vl/modelling_qwen3_vl/test_model.py /lustre/fs1/portfolios/coreai/projects/coreai_dlalgo_mcore/users/jinliangl/repos/Megatron-Bridge/src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/text_model.py`

Note: pytest was run from `/tmp` because this checkout already has `nemo_experiments/` in the repo root and the global test fixture refuses to run when that directory exists.

# GitHub Actions CI

External branch from wplf fork. An NVIDIA developer may need to trigger CI with `/ok to test 0b9d864f` if copy-pr-bot does not start automatically.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed Contributor guidelines
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation? Not needed; code and tests only.
- [x] Does the PR affect components that are optional to install? No.

### CODE REVIEW

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/utils.py [hunk 1]
```diff
@@ -636,7 +636,9 @@ def pack_dist_train_vision_module_output(
     deepstack_feature_lists: Sequence[torch.Tensor],
 ) -> dict[str, torch.Tensor]:
     """Concat deepstack features and final vision embeddings; shape as 3D for bridge communicator."""
-    vision_module_output_tensor = torch.cat([vision_embeds, *deepstack_feature_lists], dim=0)
+    # ``set_dist_train_input_tensors`` unpacks all leading chunks as deepstack features and the final
+    # chunk as the final vision embedding.
+    vision_module_output_tensor = torch.cat([*deepstack_feature_lists, vision_embeds], dim=0)
```
Comment (wplf) [inline src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/utils.py:641]: @shifangx see this function set_dist_train_input_tensors

**Comment by wplf** [comment]
/ok to test [9823be7](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3869/commits/9823be737f302e6faf148b08ebbec49008ccf528)

File: tests/unit_tests/models/qwen_vl/modelling_qwen3_vl/test_model.py [hunk 2]
```diff
@@ -603,6 +603,95 @@ def __call__(self, **kwargs):
...
+            packed_seq_params=packed_seq_params,
+        )
+
+        assert torch.equal(output, torch.ones(1))
+        assert torch.equal(captured["input_ids"], torch.tensor([[10, 11, 0], [20, 21, 22]]))
+        assert torch.equal(
+            language_model.last_kwargs["position_ids"],
+            torch.tensor([[[0, 1, 2, 0, 1, 2]], [[0, 1, 2, 0, 1, 2]], [[0, 1, 2, 0, 1, 2]]]),
+        )
+        assert language_model.rotary_pos_emb.is_thd_format is True
```
[the remark this answers is not in the document]
Comment (wplf) [reply tests/unit_tests/models/qwen_vl/modelling_qwen3_vl/test_model.py:693]: Added `test_forward_packed_thd_passes_explicit_moe_padding_mask` in `0b9d864f`. The test passes a nonzero token at an explicitly masked position, so it verifies the propagated `moe_padding_mask` is used instead of the `lm_input_ids.eq(0)` fallback after THD packing.

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/utils.py (hunk 1 above)
Comment (wplf) [reply src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/utils.py:641]: Confirmed. The pack order follows `set_dist_train_input_tensors`: leading chunks are deepstack features and the last chunk is final `vision_embeds`. The existing `test_pack_dist_train_vision_module_output_matches_unpack_contract` covers that contract.

File: src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/text_model.py [hunk 3]
```diff
@@ -156,7 +158,19 @@ def forward(
             rotary_pos_cos,
             rotary_pos_sin,
             sequence_len_offset,
-        ) = preproc_output[:5]
+            padding_mask,
+        ) = preproc_output[:6]
```
[the remark this answers is not in the document]
Comment (wplf) [reply src/megatron/bridge/models/qwen_vl/modelling_qwen3_vl/text_model.py:162]: Padding_mask, which is crucial for moe training,  is necessary for latest mcore version.
